### PR TITLE
[Dynamo][autograd.Function] Relax backward speculation strict mode a bit

### PR DESCRIPTION
**Pull request by yanboliang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #146742
* #146741
* __->__ #146571

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -2552,17 +2557,63 @@ def is_strict_for(v: VariableTracker):
...
+            except torch._dynamo.exc.Unsupported as e:
+                import traceback
+
+                if (
+                    len(e.real_stack) > 0
+                    and isinstance(e.real_stack[-1], traceback.FrameSummary)
+                    and e.real_stack[-1]._line is not None
+                    and ".is_contiguous()" in e.real_stack[-1]._line
+                    and e.real_stack[-1].name == "backward"
+                ):
```
Comment (yanboliang) [inline torch/_dynamo/variables/higher_order_ops.py:2581]: I’m unsure if this is the best way to specify the exception raised from an assert is_contiguous call. Alternatively, I could define a custom exception class, raise it when is_contiguous is called in the backward pass, and handle it here.

File: torch/_dynamo/config.py [hunk 2]
```diff
@@ -516,12 +513,11 @@ def default_debug_dir_root():
     "storage_offset",
     "layout",
     "data",
+    "is_neg",
+    "is_conj",
+    "is_pinned",
```
Comment (zou3519) [inline torch/_dynamo/config.py:518]: shouldn't we also ban is_contiguous?

Comment (zou3519) [reply torch/_dynamo/config.py:518]: Actually I'm confused. In the current version of the PR, why does the backward speculation fail on the is_contiguous call?

Comment (zou3519) [reply torch/_dynamo/config.py:518]: Yeah I think this PR is wrong. We should still ban is_contiguous here.

Comment (yanboliang) [reply torch/_dynamo/config.py:518]: Oh, the current logic does not ban `is_contiguous`; instead, we assume `grad_output` is valid and speculate the backward pass from there. If that approach fails, we catch the exception and retry with a rewritten backward pass with inserting `grad_output.contiguous()`.

From my understanding, your idea is to explicitly ban `is_contiguous`, catch the strict mode exception when it’s called, and then speculate the rewritten backward pass. That approach seems safer, and I agree with it.

File: torch/_dynamo/variables/higher_order_ops.py [hunk 3]
```diff
@@ -2552,17 +2557,66 @@ def is_strict_for(v: VariableTracker):
...
+                    "autograd.Function",
+                    enable_grad=False,
+                    set_subgraph_inputs="manual",
+                    restore_side_effects=False,
+                    tracer=bwd_tracer,
+                )
+            except torch._dynamo.exc.Unsupported as e:
+                from unittest import mock
+
+                if e.msg == "Illegal getattr invocation is_contiguous in strict mode":
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:2575]: Can you make the Illegalgetattrinvocation a specific exception type that we can handle in an except block here?

Also, it's not just is_contiguous, we can also handle .stride() calls this way.

File: torch/_dynamo/config.py [hunk 4]
```diff
@@ -516,12 +513,12 @@ def default_debug_dir_root():
...
+    "is_contiguous",
```
Comment (zou3519) [inline torch/_dynamo/config.py:519]: Let's try to do the following in some follow ups:
- we can delete .data (I don't know why that's here)
- we can delete requires_grad (long story, but AOTAutograd doesn't support create_graph=True, so the gradients never have requires_grad=True. HOWEVER we should check to see the "example" gradients that we're passing actually have requires_grad=False (if they don't then they need an extra .detach() call)

File: torch/_dynamo/variables/higher_order_ops.py [hunk 5]
```diff
@@ -2552,17 +2557,62 @@ def is_strict_for(v: VariableTracker):
...
+                    bwd_args,
+                    kwargs,
+                    "autograd.Function",
+                    enable_grad=False,
+                    set_subgraph_inputs="manual",
+                    restore_side_effects=False,
+                    tracer=bwd_tracer,
+                )
+            except torch._dynamo.exc.Unsupported as e:
+                if isinstance(e, torch._dynamo.exc.IllegalGetAttrInvocation):
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:2573]: Name this some thing like... UnknownPropertiesDuringBackwardTrace or something

**Review by zou3519 — APPROVED** [review]
